docs/language/learn-ql/ql-etudes/river-crossing.ql

```diff
@@ -0,0 +1,121 @@
+ *
+ * Note: Parts of this QL file are included in the corresponding .rst file.
+ * Make sure to update the line numbers if you change anything here!
+ */
+
+/** A possible cargo item. */
+class Cargo extends string {
+  Cargo() {
+    this = "Nothing" or
+    this = "Goat" or
+    this = "Cabbage" or
+    this = "Wolf"
+  }
+}
+
+/** One of two shores. */
+class Shore extends string {
+  Shore() {
+    this = "Left" or
+    this = "Right"
+  }
+
+  /** Returns the other shore. */
+  Shore other() {
+    this = "Left" and result = "Right"
+    or
+    this = "Right" and result = "Left"
+  }
+}
+
+/** Renders the state as a string. */
+string renderState(Shore manShore, Shore goatShore, Shore cabbageShore, Shore wolfShore) {
+  result = manShore + "," + goatShore + "," + cabbageShore + "," + wolfShore
+}
+
+/** A record of where everything is. */
+class State extends string {
+  Shore manShore;
+
+  Shore goatShore;
+
+  Shore cabbageShore;
+
+  Shore wolfShore;
+
+  State() { this = renderState(manShore, goatShore, cabbageShore, wolfShore) }
+
+  /** Returns the state that is reached after ferrying a particular cargo item. */
+  State ferry(Cargo cargo) {
+    cargo = "Nothing" and
+    result = renderState(manShore.other(), goatShore, cabbageShore, wolfShore)
+    or
+    cargo = "Goat" and
+    result = renderState(manShore.other(), goatShore.other(), cabbageShore, wolfShore)
+    or
+    cargo = "Cabbage" and
+    result = renderState(manShore.other(), goatShore, cabbageShore.other(), wolfShore)
+    or
+    cargo = "Wolf" and
+    result = renderState(manShore.other(), goatShore, cabbageShore, wolfShore.other())
+  }
+
+  /**
+   * Holds if eating occurs. This happens when predator and prey are on the same shore
+   * and the man is not present.
+   */
+  predicate eating(Shore predatorShore, Shore preyShore) {
+    predatorShore = preyShore and manShore != predatorShore
+  }
+
+  /** Holds if nothing gets eaten in this state. */
+  predicate isSafe() { not (eating(goatShore, cabbageShore) or eating(wolfShore, goatShore)) }
+
+  /** Returns the state that is reached after safely ferrying a cargo item. */
+  State safeFerry(Cargo cargo) { result = this.ferry(cargo) and result.isSafe() }
+
+  /**
+   * Returns all states that are reachable via safe ferrying.
+   * `path` keeps track of how it is achieved.
+   * `visitedStates` keeps track of previously visited states and is used to avoid loops.
+   */
+  State reachesVia(string path, string visitedStates) {
+    // Trivial case: a state is always reachable from itself.
+    this = result and
+    visitedStates = "" and
+    path = ""
+    or
+    // A state is reachable using pathSoFar and then safely ferrying cargo.
+    exists(string pathSoFar, string visitedStatesSoFar, Cargo cargo |
+      result = this.reachesVia(pathSoFar, visitedStatesSoFar).safeFerry(cargo) and
+      // The resulting state has not yet been visited.
+      not exists(int i | i = visitedStatesSoFar.indexOf(result)) and
```

Comment:
I think this could be written as `not exists(visitedStatesSoFar.indexOf(result))` and omit the variable declaration, though I'll leave it up to you to decide if that's a improvement.